src/main/java/org/wildfly/naming/client/util/NamingUtils.java

```diff
@@ -57,7 +57,7 @@ public static CompositeName toCompositeName(final Name name) throws InvalidNameE
             return (CompositeName) name;
         } else {
             final CompositeName compositeName = new CompositeName();
-            compositeName.add(name.toString());
+            compositeName.addAll(new CompositeName(name.toString()));
```

Comment:
I should qualify that, that's not the primary reason, it's just part of it.  Really what's happening is, any name can be looked up remotely, including truly composite names (like names where you have a SimpleName segment, an LDAP segment, and, say, a DNS segment (to contrive an example)).  Each of these names might use different parsers or whatever; JNDI federation is used to step through each segment and resolve it to a Context which can (hopefully) parse and resolve the subsequent segments.

Because a CompositeName *can* be passed in, I've gone and made every lookup name be a CompositeName.

My thought was that on the server, if the protocol version is 2+, we treat the name as a composite name and then parse out each segment using standard federation techniques.  I originally thought we could change the server contexts to implement org.wildfly.naming.client.AbstractFederatingContext but that may not be practical in the end.